docs/proposals/2841-flux-hpc/README.md

```diff
@@ -0,0 +1,275 @@
+
+By integrating Flux Framework, a next-generation workload manager, we can address these gaps directly. Flux provides a graph-based scheduling model and a robust `zeromq`-based bootstrapping mechanism, offering a superior alternative for demanding distributed jobs.
+
+This KEP proposes a design to integrate Flux into Kubeflow Trainer by extending the `JobSet` backend, providing a seamless user experience for running HPC-style workloads on Kubernetes.
+
+### Goals
+
+1.  **Integrate Flux Framework into Kubeflow Trainer** by creating a new plugin that extends the `JobSet` backend. This plugin will dynamically inject and configure a Flux cluster "on-the-fly" for a given `TrainJob`. While the Flux Operator [MiniCluster](https://github.com/flux-framework/flux-operator) will not be used directly, it's design strategy will be.
+2.  **Provide a Robust, SSH-less MPI Bootstrap Mechanism.** Enable users to run any MPI-enabled application without the overhead of configuring SSH keys or a dedicated MPI user, simply by leveraging Flux.
+3.  **Expose Advanced HPC Scheduling Capabilities.** Lay the groundwork for users to leverage Flux's features, such as fine-grained resource mapping, reservations, hierarchical management, and job queueing, within their Kubeflow training jobs.
+4.  **Introduce an Extensible API Policy.** Define a new `hpcPolicy` that is generic enough to support other HPC workload managers in the future (e.g., Slurm, LSF), with Flux as the initial, reference implementation.
+
+### Non-Goals
+
+1.  **Replace JobSet:** This proposal extends and enhances `JobSet`, not replaces it. `JobSet` remains the core abstraction for managing the group of pods.
+2.  **Support Additional HPC Schedulers Immediately:** The initial implementation will focus exclusively on Flux Framework. Support for other managers can be added later under the same `hpcPolicy` API by respective parties that have interest.
+3.  **Re-implement the MPI Operator:** This proposal provides an alternative to the MPI Operator's launcher/worker model by leveraging Flux's native capabilities, rather than replicating its logic.
+
+## Proposal
+
+The core of this proposal is to introduce a new Kubeflow Trainer plugin named `Flux`. This plugin will implement the `ComponentBuilderPlugin` interface to modify the `JobSet` specification generated for a `TrainJob`. The mechanism for creating the Flux MiniCluster is dynamic and non-intrusive to the user's container image:
+
+1.  **API Trigger**: The user enables the feature by defining an `hpcPolicy` in their `TrainJob` runtime specification and setting the `manager` to `"flux"`.
+2.  **Plugin Activation**: The Kubeflow Trainer controller invokes the `Flux` plugin's `Build` method.
+3.  **JobSet Modification**: The plugin modifies the `JobSet` specification before it is created:
+    *   An **`initContainer`** is added to the "trainer" replicated job. This container uses a pre-built "flux-view" image containing a Spack installation of Flux.
```

Comment:
Do you require launcher + node rJobs for Flux, or we can have single rJob like in torch-distributed: https://github.com/kubeflow/trainer/blob/master/manifests/base/runtimes/torch_distributed.yaml

Reply:
@vsoch Which entrypoint we will use to trigger the task? Will it still be `mpirun` ?

Reply:
`mpirun` is an analogous MPI launcher. Flux bootstraps MPI, so the entrypoint is `flux run` or `flux submit` (from inside the MiniCluster). You would use one _or_ the other but not both.